app/workers/email_generation_worker.rb

```diff
@@ -1,24 +1,38 @@
 class EmailGenerationWorker
   include Sidekiq::Worker
 
-  def perform(subscription_content_id)
-    subscription_content = SubscriptionContent
-      .includes(:content_change)
-      .includes(subscription: :subscriber)
-      .find(subscription_content_id)
-
-    email = Email.create_from_params!(email_params(subscription_content))
+  def perform
+    while subscription_contents.exists?
+      to_queue = []
 
-    subscription_content.update!(email: email)
+      SubscriptionContent.transaction do
+        subscription_contents.lock.limit(100).find_each do |subscription_content|
+          email = create_email(subscription_content: subscription_content)
+          subscription_content.update!(email: email)
```

Comment:
My understanding is we need to handle this race condition:

- Two of these jobs run at the same time
- Each job creates an `email` for a single `subscription_content`
- The last job 'wins' the race and sets the `subscription_content.email_id` field
- The other email is 'orphaned' and in an inconsistent state

I think your approach of pessimistically locking the table works, but maybe there's a way to do this optimistically:

- Add a unique index to the `subscription_content.email_id` field in the database
- Create the `email` and associate the `subscription_content` with it in a transaction
- Rescue from this operation if we get a `UniqueViolation` and return from the job

This way, we assume there aren't two processes creating/updating the database at the same time and if there are, the first process makes the change and the second returns from the job, doing nothing.

However, the overhead of transactionalising each create/update might add some overhead. We should measure this.

Reply:
I like that solution and I think it should work - we may have to put in a `return if subscription_content.email` somewhere to avoid sending out duplicate emails. I'll leave this discussion open to see what other people think for the moment.

Reply:
We could set `success=true` after the `#update!` line and exit the job early if this isn't set, maybe?